aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,911 @@
+///
+/// It is strongly discouraged to:
+///   - Delete active or expired keypairs too early, as that would leave Janus unable to decrypt
+///     report shares using the keypair.
+///   - Promote pending keypairs too early, unless all Janus replicas have been rebooted since
+///     their insertion.
+///   - Directly insert active keypairs.
+///
+/// Keypairs that are manually inserted are adopted by the key rotator and will have their lifecycle
+/// managed. The number of keypairs per ciphersuite will trend towards 1, i.e. 2 keys with the same
+/// ciphersuite will eventually be replaced with 1 key.
+#[derive(Debug)]
+pub struct KeyRotator<C: Clock> {
+    datastore: Arc<Datastore<C>>,
+    hpke: Vec<HpkeKeyRotatorConfig>,
+}
+
+/// Defines the ciphersuite and rotation policy of a global HPKE key.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct HpkeKeyRotatorConfig {
+    /// How long key remains in [`HpkeKeyState::Pending`] before being moved to
+    /// [`HpkeKeyState::Active`]. This should be greater than
+    /// [`AggregatorConfig::global_hpke_configs_refresh_interval`].
+    #[serde(rename = "pending_duration_secs", default = "default_pending_duration")]
+    pub pending_duration: Duration,
+
+    /// The time-to-live of the key. Once this is exceeded, the key is moved to
+    /// [`HpkeKeyState::Expired`]. It is at operator discretion as to how long this should be.
+    #[serde(rename = "active_duration_secs", default = "default_active_duration")]
+    pub active_duration: Duration,
+
+    /// How long the key remains in [`HpkeKeyState::Expired`] before being deleted. This should
+    /// be greater than the clients' HPKE key cache maximum age.
+    #[serde(rename = "expired_duration_secs", default = "default_expired_duration")]
+    pub expired_duration: Duration,
+
+    /// The ciphersuite of the key.
+    #[serde(default = "default_hpke_ciphersuite")]
+    pub ciphersuite: HpkeCiphersuite,
+
+    /// Safely phase out this ciphersuite. This will immediately delete any pending keys and
+    /// expire any active keys. Expired keys are retained for at least [`Self::expired_duration`]
+    /// before being deleted.
+    ///
+    /// After [`Self::expired_duration`], this config can be removed entirely. It is not recommended
+    /// to remove key configurations without running the key rotator with `retire` first.
+    ///
+    /// There must be at least one non-retired config present with its key in the active state. If
+    /// this is the only config and it is marked retired, the key rotator will refuse to run.
+    #[serde(default)]
+    pub retire: bool,
+}
+
+/// Returns [`GlobalHpkeKeypairCache::DEFAULT_REFRESH_INTERVAL`] times 2, for safety margin.
+fn default_pending_duration() -> Duration {
+    Duration::from_seconds(GlobalHpkeKeypairCache::DEFAULT_REFRESH_INTERVAL.as_secs() * 2)
+}
+
+/// 4 weeks. This is long enough not to be unnecessary churn, but short enough that misbehaving
```

Comment:
I was hoping that a short rotation would force clients to have proper key fetching and caching strategy, i.e. the surprises would happen in the 1 month timespan rather than the 1 year timespan, and we would discourage any permanent retention of the key.

Short rotation also validates whether all this stuff works in production, so that we're not shocked 1 year later when we've completely forgotten how this tool works.

Reply:
That's fair, but I do feel strongly that 1 year is too long, for at least the other reasons. Can split the difference though, 3 months? 6 months?

> Still, if that's not convincing, let's make sure we communicate with all deployed folks before we enable key rotation in their respective environments.

:+1: